File: webrtc/src/main/java/com/bandwidth/webrtc/signaling/Signaling.java
package com.bandwidth.webrtc.signaling;

import com.bandwidth.webrtc.signaling.listeners.OnConnectListener;
import com.bandwidth.webrtc.signaling.listeners.OnDisconnectListener;
import com.bandwidth.webrtc.signaling.rpc.transit.OfferSdpResult;
import com.bandwidth.webrtc.signaling.rpc.transit.RequestToPublishResult;

import java.net.URI;

public interface Signaling {
    void setOnConnectListener(OnConnectListener listener);
    void setOnDisconnectListener(OnDisconnectListener listener);

    void connect(URI uri) throws ConnectionException;
    void disconnect();
    void sendIceCandidate(String endpointId, String sdp, Integer sdpMLineIndex, String sdpMid);

    void offerSdp(String endpointId, String sdp, Observer observer);
    void requestToPublish(Boolean audio, Boolean video, String alias, Observer observer);
    void setMediaPreferences(Observer observer);

    interface Observer {
        void onOfferSdp(Signaling signaling, OfferSdpResult result);
        void onRequestToPublish(Signaling signaling, RequestToPublishResult result);
        void onSetMediaPreferences(Signaling signaling);
    }

    class Adapter implements Observer {
        @Override
        public void onOfferSdp(Signaling signaling, OfferSdpResult result) {

        }

        @Override
        public void onRequestToPublish(Signaling signaling, RequestToPublishResult result) {

        }

        @Override
        public void onSetMediaPreferences(Signaling signaling) {

        }
    }
}

File: webrtc/src/main/java/com/bandwidth/webrtc/signaling/Signaling.kt
package com.bandwidth.webrtc.signaling

import com.bandwidth.webrtc.signaling.rpc.transit.OfferSdpResult
import com.bandwidth.webrtc.types.PublishMetadata
import com.bandwidth.webrtc.signaling.listeners.OnDisconnectListener
import com.bandwidth.webrtc.signaling.rpc.transit.AnswerSdpResult
import com.bandwidth.webrtc.signaling.rpc.transit.SetMediaPreferencesResult

interface Signaling {
    fun setOnConnectListener(listener: (Signaling?) -> Unit)
    fun setOnDisconnectListener(listener: OnDisconnectListener?)

    @Throws(ConnectionException::class)
    fun connect(deviceToken: String?)

    @Throws(ConnectionException::class)
    fun connect(webSocketUrl: String?, deviceToken: String?)
    fun disconnect()
    fun offerSdp(sdp: String, publishMetadata: PublishMetadata, observer: Observer)
    fun answerSdp(sdp: String, observer: Observer)
    interface Observer {
        fun onOfferSdp(signaling: Signaling?, result: OfferSdpResult?)
        fun onAnswerSdp(signaling: Signaling?, result: AnswerSdpResult?)
        fun onSetMediaPreferences(signaling: Signaling?, result: SetMediaPreferencesResult?)
    }

    open class Adapter : Observer {
        override fun onOfferSdp(signaling: Signaling?, result: OfferSdpResult?) {}
        override fun onAnswerSdp(signaling: Signaling?, result: AnswerSdpResult?) {}
        override fun onSetMediaPreferences(
            signaling: Signaling?,
            result: SetMediaPreferencesResult?
        ) {
        }
    }
}